proposals/pr002-hunkspace.md

```diff
@@ -0,0 +1,543 @@
+these worlds get to interact at some level. For example Firefox L10N can not
+exist in its current extensible form on Fuchsia, due to the tight controls that
+the OS puts in place on the runtime. (You'd have to download a prebuilt version
+for your locale, from Mozilla.)  Having a cross-platform data provider
+specification could make that a non-issue.]
+
+## Unique IDs
+
+Every localizable resource should get a unique identifier.  This allows
+a platform-level localization clearinghouse (from the example above) to
+be a one-stop data provider for localization needs.
+
+For example, the [current proposal of the icu4x data provider for
+CLDR][dpcldr] uses a 4-dimensional key space consisting of:
+
+- A Language identifier;
+- a Category;
+- a Key;
+- a Payload.
+
+[dpcldr]: https://github.com/unicode-org/icu4x/blob/master/components/data-provider/src/lib.rs#L52
+
+This key space is intended to fit any CLDR resource according to the data
+mapping which is internally defined by CLDR rules.
+
+It does not represent a fictitious resource string "Hello, {user}" from
+Firefox v143, for example.  So we allow the key space to have more
+dimensions than the current 4 of CLDR, to allow us to address both
+resources uniquely.  Call that key space a "hunkspace".
+
+At minimum this means adding a "source" dimension to the key space,
+allowing for example "cldr.unicode.org" to be used as a (fixed)
+dimension for all CLDR data.  We could have a "fuchsia.com" source that
+could use a custom data schema like so:
+
+```
+Key: {
+  source = "fuchsia.dev",
+  app="weirdy-birds",
+  message_id=42,
+  locale="en-US"
+}
+Value: "Hello, {user}"
+```
+
+
+A note is in order on what constitutes "identical" messages.
+Two messages are identical if and only if they map to the same point in
+the hunkspace, which means that they are interchangeable no matter what
+the context.
+
+Two messages that happen to have identical strings, such as "Print" (1)
+and "Print" (2) won't necessarily be interchangeable, if one is used to
+mean "Go print this!", while the other is used to stand for "something
+which is printed".  This is because these two messages could be
+displayed differently in other languages. "Штампај", "Одштампано" would
+be the respective translations in another language. The hope here is
+that it would always be possible to split a keyspace point into two by
+providing finer-grained context where needed, but discussions of that
+are out of scope.
+
+A CLDR-only data provider would be a data provider which fixes the
+"source" string to the value "cldr.unicode.org".
+
+## Requirements
+Some requirements distilled from the discussion above:
+
+1.  There must be a commonly understood data schema for each data set.
+     For example "CLDR unicode data",
+2.  It must be possible for that data to come from multiple sources.
+     E.g. "mozilla.org", vs "google.com", vs. "unicode.org".  Every
+    source that provides a specific data schema can have own data.
```

Comment:
The data sources for "CLDR unicode data" should provide the data hunks in the same schema, so that consumers can read and use the data using the same code.

Reply:
This should be supported by say cldr://unicode.org/...